`storage.writer(...)` does not automatically retry unexpected EOF when parsing json for an idempotent session.

artifact: `com.google.cloud:google-cloud-storage:2.2.1`

#### EOF @ offset 0
When the unexpected EOF at the 0 byte an `java.lang.IllegalArgumentException` is encountered

<details><summary>Stacktrace</summary>
<p>

```
com.google.cloud.storage.StorageException: java.lang.IllegalArgumentException: no JSON input found
	at com.google.cloud.storage.StorageException.getStorageException(StorageException.java:85)
	at com.google.cloud.storage.StorageException.translateAndThrow(StorageException.java:81)
	at com.google.cloud.storage.BlobWriteChannel.flushBuffer(BlobWriteChannel.java:275)
	at com.google.cloud.BaseWriteChannel.close(BaseWriteChannel.java:151)
	at com.google.cloud.storage.it.ITBlobWriteChannelTest.doJsonUnexpectedEOFTest(ITBlobWriteChannelTest.java:185)
	at com.google.cloud.storage.it.ITBlobWriteChannelTest.testJsonEOF_0B(ITBlobWriteChannelTest.java:102)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.google.cloud.storage.DataGeneration$1.evaluate(DataGeneration.java:56)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at com.google.cloud.storage.conformance.retry.TestBench$1.evaluate(TestBench.java:232)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:69)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:33)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:235)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:54)
Caused by: java.lang.IllegalArgumentException: no JSON input found
	at com.google.common.base.Preconditions.checkArgument(Preconditions.java:145)
	at com.google.api.client.util.Preconditions.checkArgument(Preconditions.java:47)
	at com.google.api.client.json.JsonParser.startParsing(JsonParser.java:215)
	at com.google.api.client.json.JsonParser.parse(JsonParser.java:358)
	at com.google.api.client.json.JsonParser.parse(JsonParser.java:335)
	at com.google.api.client.json.JsonObjectParser.parseAndClose(JsonObjectParser.java:79)
	at com.google.api.client.json.JsonObjectParser.parseAndClose(JsonObjectParser.java:73)
	at com.google.api.client.http.HttpResponse.parseAs(HttpResponse.java:461)
	at com.google.cloud.storage.spi.v1.HttpStorageRpc.writeWithResponse(HttpStorageRpc.java:882)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.google.cloud.storage.it.ITBlobWriteChannelTest.lambda$null$0(ITBlobWriteChannelTest.java:161)
	at com.sun.proxy.$Proxy7.writeWithResponse(Unknown Source)
	at com.google.cloud.storage.BlobWriteChannel.transmitChunk(BlobWriteChannel.java:67)
	at com.google.cloud.storage.BlobWriteChannel.access$1400(BlobWriteChannel.java:34)
	at com.google.cloud.storage.BlobWriteChannel$1.run(BlobWriteChannel.java:252)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at com.google.api.gax.retrying.DirectRetryingExecutor.submit(DirectRetryingExecutor.java:105)
	at com.google.cloud.RetryHelper.run(RetryHelper.java:76)
	at com.google.cloud.RetryHelper.runWithRetries(RetryHelper.java:50)
	at com.google.cloud.storage.BlobWriteChannel.flushBuffer(BlobWriteChannel.java:183)
	... 32 more
```
</p>
</details>

#### EOF @ offset > 0
When the unexpected EOF at an offset > 0 byte a `com.fasterxml.jackson.core.io.JsonEOFException` is encountered

<details><summary>Stacktrace</summary>
<p>

```
com.google.cloud.storage.StorageException: Unexpected end-of-input in VALUE_STRING
 at [Source: (BufferedInputStream); line: 1, column: 11]
	at com.google.cloud.storage.spi.v1.HttpStorageRpc.translate(HttpStorageRpc.java:233)
	at com.google.cloud.storage.spi.v1.HttpStorageRpc.writeWithResponse(HttpStorageRpc.java:901)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.google.cloud.storage.it.ITBlobWriteChannelTest.lambda$null$0(ITBlobWriteChannelTest.java:161)
	at com.sun.proxy.$Proxy7.writeWithResponse(Unknown Source)
	at com.google.cloud.storage.BlobWriteChannel.transmitChunk(BlobWriteChannel.java:67)
	at com.google.cloud.storage.BlobWriteChannel.access$1400(BlobWriteChannel.java:34)
	at com.google.cloud.storage.BlobWriteChannel$1.run(BlobWriteChannel.java:252)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at com.google.api.gax.retrying.DirectRetryingExecutor.submit(DirectRetryingExecutor.java:105)
	at com.google.cloud.RetryHelper.run(RetryHelper.java:76)
	at com.google.cloud.RetryHelper.runWithRetries(RetryHelper.java:50)
	at com.google.cloud.storage.BlobWriteChannel.flushBuffer(BlobWriteChannel.java:183)
	at com.google.cloud.BaseWriteChannel.close(BaseWriteChannel.java:151)
	at com.google.cloud.storage.it.ITBlobWriteChannelTest.doJsonUnexpectedEOFTest(ITBlobWriteChannelTest.java:185)
	at com.google.cloud.storage.it.ITBlobWriteChannelTest.testJsonEOF_10B(ITBlobWriteChannelTest.java:113)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.google.cloud.storage.DataGeneration$1.evaluate(DataGeneration.java:56)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at com.google.cloud.storage.conformance.retry.TestBench$1.evaluate(TestBench.java:232)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:69)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:33)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:235)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:54)
Caused by: com.fasterxml.jackson.core.io.JsonEOFException: Unexpected end-of-input in VALUE_STRING
 at [Source: (BufferedInputStream); line: 1, column: 11]
	at com.fasterxml.jackson.core.base.ParserMinimalBase._reportInvalidEOF(ParserMinimalBase.java:682)
	at com.fasterxml.jackson.core.base.ParserMinimalBase._reportInvalidEOF(ParserMinimalBase.java:663)
	at com.fasterxml.jackson.core.json.UTF8StreamJsonParser._loadMoreGuaranteed(UTF8StreamJsonParser.java:2444)
	at com.fasterxml.jackson.core.json.UTF8StreamJsonParser._finishAndReturnString(UTF8StreamJsonParser.java:2485)
	at com.fasterxml.jackson.core.json.UTF8StreamJsonParser.getText(UTF8StreamJsonParser.java:334)
	at com.google.api.client.json.jackson2.JacksonParser.getText(JacksonParser.java:74)
	at com.google.api.client.json.JsonParser.parseValue(JsonParser.java:856)
	at com.google.api.client.json.JsonParser.parse(JsonParser.java:451)
	at com.google.api.client.json.JsonParser.parseValue(JsonParser.java:787)
	at com.google.api.client.json.JsonParser.parse(JsonParser.java:360)
	at com.google.api.client.json.JsonParser.parse(JsonParser.java:335)
	at com.google.api.client.json.JsonObjectParser.parseAndClose(JsonObjectParser.java:79)
	at com.google.api.client.json.JsonObjectParser.parseAndClose(JsonObjectParser.java:73)
	at com.google.api.client.http.HttpResponse.parseAs(HttpResponse.java:461)
	at com.google.cloud.storage.spi.v1.HttpStorageRpc.writeWithResponse(HttpStorageRpc.java:882)
	... 46 more
```
</p>
</details>

--- a/google-cloud-storage/pom.xml
+++ b/google-cloud-storage/pom.xml
@@ -87,6 +87,11 @@
       <groupId>org.threeten</groupId>
       <artifactId>threetenbp</artifactId>
     </dependency>
+    <!-- Access to exception for retry handling -->
+    <dependency>
+      <groupId>com.fasterxml.jackson.core</groupId>
+      <artifactId>jackson-core</artifactId>
+    </dependency>
 
     <!-- Test dependencies -->
     <dependency>

--- a/google-cloud-storage/src/main/java/com/google/cloud/storage/DefaultStorageRetryStrategy.java
+++ b/google-cloud-storage/src/main/java/com/google/cloud/storage/DefaultStorageRetryStrategy.java
@@ -16,11 +16,13 @@
 
 package com.google.cloud.storage;
 
+import com.fasterxml.jackson.core.io.JsonEOFException;
 import com.google.api.client.http.HttpResponseException;
 import com.google.cloud.BaseServiceException;
 import com.google.cloud.ExceptionHandler;
 import com.google.cloud.ExceptionHandler.Interceptor;
 import com.google.common.collect.ImmutableSet;
+import com.google.gson.stream.MalformedJsonException;
 import java.io.IOException;
 import java.util.Set;
 
@@ -33,7 +35,8 @@ final class DefaultStorageRetryStrategy implements StorageRetryStrategy {
   private static final Interceptor INTERCEPTOR_NON_IDEMPOTENT =
       new InterceptorImpl(false, ImmutableSet.of());
 
-  private static final ExceptionHandler IDEMPOTENT_HANDLER = newHandler(INTERCEPTOR_IDEMPOTENT);
+  private static final ExceptionHandler IDEMPOTENT_HANDLER =
+      newHandler(new EmptyJsonParsingExceptionInterceptor(), INTERCEPTOR_IDEMPOTENT);
   private static final ExceptionHandler NON_IDEMPOTENT_HANDLER =
       newHandler(INTERCEPTOR_NON_IDEMPOTENT);
 
@@ -47,15 +50,11 @@ final class DefaultStorageRetryStrategy implements StorageRetryStrategy {
     return NON_IDEMPOTENT_HANDLER;
   }
 
-  private static ExceptionHandler newHandler(Interceptor interceptor) {
-    return ExceptionHandler.newBuilder()
-        .retryOn(BaseServiceException.class)
-        .retryOn(IOException.class)
-        .addInterceptors(interceptor)
-        .build();
+  private static ExceptionHandler newHandler(Interceptor... interceptors) {
+    return ExceptionHandler.newBuilder().addInterceptors(interceptors).build();
   }
 
-  private static class InterceptorImpl implements Interceptor {
+  private static class InterceptorImpl implements BaseInterceptor {
 
     private static final long serialVersionUID = -5153236691367895096L;
     private final boolean idempotent;
@@ -64,11 +63,6 @@ final class DefaultStorageRetryStrategy implements StorageRetryStrategy {
     private InterceptorImpl(boolean idempotent, Set<BaseServiceException.Error> retryableErrors) {
       this.idempotent = idempotent;
       this.retryableErrors = ImmutableSet.copyOf(retryableErrors);
-    }
-
-    @Override
-    public RetryResult afterEval(Exception exception, RetryResult retryResult) {
-      return RetryResult.CONTINUE_EVALUATION;
     }
 
     @Override
@@ -95,7 +89,11 @@ final class DefaultStorageRetryStrategy implements StorageRetryStrategy {
     }
 
     private RetryResult shouldRetryIOException(IOException ioException) {
-      if (BaseServiceException.isRetryable(idempotent, ioException)) {
+      if (ioException instanceof JsonEOFException && idempotent) { // Jackson
+        return RetryResult.RETRY;
+      } else if (ioException instanceof MalformedJsonException && idempotent) { // Gson
+        return RetryResult.RETRY;
+      } else if (BaseServiceException.isRetryable(idempotent, ioException)) {
         return RetryResult.RETRY;
       } else {
         return RetryResult.NO_RETRY;
@@ -115,6 +113,28 @@ final class DefaultStorageRetryStrategy implements StorageRetryStrategy {
       int code = baseServiceException.getCode();
       String reason = baseServiceException.getReason();
       return shouldRetryCodeReason(code, reason);
+    }
+  }
+
+  private static final class EmptyJsonParsingExceptionInterceptor implements BaseInterceptor {
+    private static final long serialVersionUID = -3320984020388043628L;
+
+    @Override
+    public RetryResult beforeEval(Exception exception) {
+      if (exception instanceof IllegalArgumentException) {
+        IllegalArgumentException illegalArgumentException = (IllegalArgumentException) exception;
+        if (illegalArgumentException.getMessage().equals("no JSON input found")) {
+          return RetryResult.RETRY;
+        }
+      }
+      return RetryResult.CONTINUE_EVALUATION;
+    }
+  }
+
+  private interface BaseInterceptor extends Interceptor {
+    @Override
+    default RetryResult afterEval(Exception exception, RetryResult retryResult) {
+      return RetryResult.CONTINUE_EVALUATION;
     }
   }
 }

--- a/google-cloud-storage/src/test/java/com/google/cloud/storage/DataGeneration.java
+++ b/google-cloud-storage/src/test/java/com/google/cloud/storage/DataGeneration.java
@@ -1,0 +1,60 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.storage;
+
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.Random;
+import org.junit.rules.TestRule;
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+
+public final class DataGeneration implements TestRule {
+
+  private final Random rand;
+
+  public DataGeneration(Random rand) {
+    this.rand = rand;
+  }
+
+  public ByteBuffer randByteBuffer(int limit) {
+    ByteBuffer b = ByteBuffer.allocate(limit);
+    fillByteBuffer(b);
+    return b;
+  }
+
+  public void fillByteBuffer(ByteBuffer b) {
+    while (b.position() < b.limit()) {
+      int i = rand.nextInt('z');
+      char c = (char) i;
+      if (Character.isLetter(c) || Character.isDigit(c)) {
+        b.put(Character.toString(c).getBytes(StandardCharsets.UTF_8));
+      }
+    }
+    b.position(0);
+  }
+
+  @Override
+  public Statement apply(Statement base, Description description) {
+    return new Statement() {
+      @Override
+      public void evaluate() throws Throwable {
+        base.evaluate();
+      }
+    };
+  }
+}

--- a/google-cloud-storage/src/test/java/com/google/cloud/storage/DefaultRetryHandlingBehaviorTest.java
+++ b/google-cloud-storage/src/test/java/com/google/cloud/storage/DefaultRetryHandlingBehaviorTest.java
@@ -20,6 +20,8 @@ import static com.google.common.collect.ImmutableList.toImmutableList;
 import static com.google.common.collect.ImmutableSet.toImmutableSet;
 import static com.google.common.truth.Truth.assertWithMessage;
 
+import com.fasterxml.jackson.core.JsonToken;
+import com.fasterxml.jackson.core.io.JsonEOFException;
 import com.google.api.client.googleapis.json.GoogleJsonError;
 import com.google.api.client.http.HttpHeaders;
 import com.google.api.client.http.HttpResponseException;
@@ -27,6 +29,7 @@ import com.google.api.gax.retrying.ResultRetryAlgorithm;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableSet;
 import com.google.common.collect.Sets;
+import com.google.gson.stream.MalformedJsonException;
 import java.io.IOException;
 import java.net.SocketException;
 import java.net.SocketTimeoutException;
@@ -319,6 +322,14 @@ public final class DefaultRetryHandlingBehaviorTest {
             "connectionClosedPrematurely",
             "connectionClosedPrematurely",
             C.CONNECTION_CLOSED_PREMATURELY)),
+    EMPTY_JSON_PARSE_ERROR(new IllegalArgumentException("no JSON input found")),
+    JACKSON_EOF_EXCEPTION(C.JACKSON_EOF_EXCEPTION),
+    STORAGE_EXCEPTION_0_JACKSON_EOF_EXCEPTION(
+        new StorageException(0, "parse error", C.JACKSON_EOF_EXCEPTION)),
+    GSON_MALFORMED_EXCEPTION(C.GSON_MALFORMED_EXCEPTION),
+    STORAGE_EXCEPTION_0_GSON_MALFORMED_EXCEPTION(
+        new StorageException(0, "parse error", C.GSON_MALFORMED_EXCEPTION)),
+    IO_EXCEPTION(new IOException("no retry")),
     ;
 
     private final Throwable throwable;
@@ -385,6 +396,10 @@ public final class DefaultRetryHandlingBehaviorTest {
           new IllegalArgumentException("illegal argument");
       private static final IOException CONNECTION_CLOSED_PREMATURELY =
           new IOException("simulated Connection closed prematurely");
+      private static final JsonEOFException JACKSON_EOF_EXCEPTION =
+          new JsonEOFException(null, JsonToken.VALUE_STRING, "parse-exception");
+      private static final MalformedJsonException GSON_MALFORMED_EXCEPTION =
+          new MalformedJsonException("parse-exception");
 
       private static HttpResponseException newHttpResponseException(
           int httpStatusCode, String name) {
@@ -913,7 +928,67 @@ public final class DefaultRetryHandlingBehaviorTest {
                 ThrowableCategory.STORAGE_EXCEPTION_0_INTERNAL_ERROR,
                 HandlerCategory.NONIDEMPOTENT,
                 ExpectRetry.NO,
-                Behavior.DEFAULT_MORE_STRICT))
+                Behavior.DEFAULT_MORE_STRICT),
+            new Case(
+                ThrowableCategory.EMPTY_JSON_PARSE_ERROR,
+                HandlerCategory.IDEMPOTENT,
+                ExpectRetry.YES,
+                Behavior.DEFAULT_MORE_PERMISSIBLE),
+            new Case(
+                ThrowableCategory.EMPTY_JSON_PARSE_ERROR,
+                HandlerCategory.NONIDEMPOTENT,
+                ExpectRetry.NO,
+                Behavior.SAME),
+            new Case(
+                ThrowableCategory.IO_EXCEPTION,
+                HandlerCategory.IDEMPOTENT,
+                ExpectRetry.NO,
+                Behavior.SAME),
+            new Case(
+                ThrowableCategory.IO_EXCEPTION,
+                HandlerCategory.NONIDEMPOTENT,
+                ExpectRetry.NO,
+                Behavior.SAME),
+            new Case(
+                ThrowableCategory.JACKSON_EOF_EXCEPTION,
+                HandlerCategory.IDEMPOTENT,
+                ExpectRetry.YES,
+                Behavior.DEFAULT_MORE_PERMISSIBLE),
+            new Case(
+                ThrowableCategory.JACKSON_EOF_EXCEPTION,
+                HandlerCategory.NONIDEMPOTENT,
+                ExpectRetry.NO,
+                Behavior.SAME),
+            new Case(
+                ThrowableCategory.STORAGE_EXCEPTION_0_JACKSON_EOF_EXCEPTION,
+                HandlerCategory.IDEMPOTENT,
+                ExpectRetry.YES,
+                Behavior.DEFAULT_MORE_PERMISSIBLE),
+            new Case(
+                ThrowableCategory.STORAGE_EXCEPTION_0_JACKSON_EOF_EXCEPTION,
+                HandlerCategory.NONIDEMPOTENT,
+                ExpectRetry.NO,
+                Behavior.SAME),
+            new Case(
+                ThrowableCategory.GSON_MALFORMED_EXCEPTION,
+                HandlerCategory.IDEMPOTENT,
+                ExpectRetry.YES,
+                Behavior.DEFAULT_MORE_PERMISSIBLE),
+            new Case(
+                ThrowableCategory.GSON_MALFORMED_EXCEPTION,
+                HandlerCategory.NONIDEMPOTENT,
+                ExpectRetry.NO,
+                Behavior.SAME),
+            new Case(
+                ThrowableCategory.STORAGE_EXCEPTION_0_GSON_MALFORMED_EXCEPTION,
+                HandlerCategory.IDEMPOTENT,
+                ExpectRetry.YES,
+                Behavior.DEFAULT_MORE_PERMISSIBLE),
+            new Case(
+                ThrowableCategory.STORAGE_EXCEPTION_0_GSON_MALFORMED_EXCEPTION,
+                HandlerCategory.NONIDEMPOTENT,
+                ExpectRetry.NO,
+                Behavior.SAME))
         .build();
   }
 }

--- a/google-cloud-storage/src/test/java/com/google/cloud/storage/PackagePrivateMethodWorkarounds.java
+++ b/google-cloud-storage/src/test/java/com/google/cloud/storage/PackagePrivateMethodWorkarounds.java
@@ -16,7 +16,11 @@
 
 package com.google.cloud.storage;
 
+import com.google.api.services.storage.model.StorageObject;
+import com.google.cloud.WriteChannel;
 import com.google.cloud.storage.BucketInfo.BuilderImpl;
+import java.util.Optional;
+import java.util.function.Function;
 
 /**
  * Several classes in the High Level Model for storage include package-local constructors and
@@ -37,5 +41,16 @@ public final class PackagePrivateMethodWorkarounds {
   public static Blob blobCopyWithStorage(Blob b, Storage s) {
     BlobInfo.BuilderImpl builder = (BlobInfo.BuilderImpl) BlobInfo.fromPb(b.toPb()).toBuilder();
     return new Blob(s, builder);
+  }
+
+  public static Function<WriteChannel, Optional<StorageObject>> maybeGetStorageObjectFunction() {
+    return (w) -> {
+      if (w instanceof BlobWriteChannel) {
+        BlobWriteChannel blobWriteChannel = (BlobWriteChannel) w;
+        return Optional.of(blobWriteChannel.getStorageObject());
+      } else {
+        return Optional.empty();
+      }
+    };
   }
 }

--- a/google-cloud-storage/src/test/java/com/google/cloud/storage/conformance/retry/CleanupStrategy.java
+++ b/google-cloud-storage/src/test/java/com/google/cloud/storage/conformance/retry/CleanupStrategy.java
@@ -16,7 +16,7 @@
 
 package com.google.cloud.storage.conformance.retry;
 
-enum CleanupStrategy {
+public enum CleanupStrategy {
   ALWAYS,
   ONLY_ON_SUCCESS,
   NEVER

--- a/google-cloud-storage/src/test/java/com/google/cloud/storage/conformance/retry/RetryTestFixture.java
+++ b/google-cloud-storage/src/test/java/com/google/cloud/storage/conformance/retry/RetryTestFixture.java
@@ -21,15 +21,10 @@ import static org.junit.Assert.assertTrue;
 import com.google.api.gax.retrying.RetrySettings;
 import com.google.api.gax.rpc.FixedHeaderProvider;
 import com.google.cloud.NoCredentials;
-import com.google.cloud.conformance.storage.v1.InstructionList;
-import com.google.cloud.conformance.storage.v1.Method;
 import com.google.cloud.storage.Storage;
 import com.google.cloud.storage.StorageOptions;
 import com.google.cloud.storage.conformance.retry.TestBench.RetryTestResource;
 import com.google.common.collect.ImmutableMap;
-import com.google.gson.JsonArray;
-import com.google.gson.JsonObject;
-import java.util.Map;
 import java.util.logging.Logger;
 import org.junit.AssumptionViolatedException;
 import org.junit.rules.TestRule;
@@ -88,7 +83,7 @@ final class RetryTestFixture implements TestRule {
         try {
           LOGGER.fine("Setting up retry_test resource...");
           RetryTestResource retryTestResource =
-              newRetryTestResource(
+              RetryTestResource.newRetryTestResource(
                   testRetryConformance.getMethod(), testRetryConformance.getInstruction());
           retryTest = testBench.createRetryTest(retryTestResource);
           LOGGER.fine("Setting up retry_test resource complete");
@@ -123,17 +118,6 @@ final class RetryTestFixture implements TestRule {
         || ((testSuccess || testSkipped) && cleanupStrategy == CleanupStrategy.ONLY_ON_SUCCESS);
   }
 
-  private static RetryTestResource newRetryTestResource(Method m, InstructionList l) {
-    RetryTestResource resource = new RetryTestResource();
-    resource.instructions = new JsonObject();
-    JsonArray instructions = new JsonArray();
-    for (String s : l.getInstructionsList()) {
-      instructions.add(s);
-    }
-    resource.instructions.add(m.getName(), instructions);
-    return resource;
-  }
-
   private Storage newStorage(boolean forTest) {
     StorageOptions.Builder builder =
         StorageOptions.newBuilder()
@@ -145,23 +129,14 @@ final class RetryTestFixture implements TestRule {
     if (forTest) {
       builder
           .setHeaderProvider(
-              new FixedHeaderProvider() {
-                @Override
-                public Map<String, String> getHeaders() {
-                  return ImmutableMap.of(
-                      "x-retry-test-id", retryTest.id, "User-Agent", fmtUserAgent("test"));
-                }
-              })
+              FixedHeaderProvider.create(
+                  ImmutableMap.of(
+                      "x-retry-test-id", retryTest.id, "User-Agent", fmtUserAgent("test"))))
           .setRetrySettings(retrySettingsBuilder.setMaxAttempts(3).build());
     } else {
       builder
           .setHeaderProvider(
-              new FixedHeaderProvider() {
-                @Override
-                public Map<String, String> getHeaders() {
-                  return ImmutableMap.of("User-Agent", fmtUserAgent("non-test"));
-                }
-              })
+              FixedHeaderProvider.create(ImmutableMap.of("User-Agent", fmtUserAgent("non-test"))))
           .setRetrySettings(retrySettingsBuilder.setMaxAttempts(1).build());
     }
     return builder.build().getService();

--- a/google-cloud-storage/src/test/java/com/google/cloud/storage/conformance/retry/TestBench.java
+++ b/google-cloud-storage/src/test/java/com/google/cloud/storage/conformance/retry/TestBench.java
@@ -30,6 +30,8 @@ import com.google.api.core.NanoClock;
 import com.google.api.gax.retrying.BasicResultRetryAlgorithm;
 import com.google.api.gax.retrying.RetrySettings;
 import com.google.cloud.RetryHelper.RetryHelperException;
+import com.google.cloud.conformance.storage.v1.InstructionList;
+import com.google.cloud.conformance.storage.v1.Method;
 import com.google.common.collect.ImmutableList;
 import com.google.gson.Gson;
 import com.google.gson.JsonArray;
@@ -60,7 +62,7 @@ import org.threeten.bp.Duration;
  *
  * <p>This rule expects to be bound as an {@link org.junit.ClassRule @ClassRule} field.
  */
-final class TestBench implements TestRule {
+public final class TestBench implements TestRule {
 
   private static final Logger LOGGER = Logger.getLogger(TestBench.class.getName());
 
@@ -69,6 +71,7 @@ final class TestBench implements TestRule {
   private final String dockerImageName;
   private final String dockerImageTag;
   private final CleanupStrategy cleanupStrategy;
+  private final String containerName;
 
   private final Gson gson;
   private final HttpRequestFactory requestFactory;
@@ -78,12 +81,14 @@ final class TestBench implements TestRule {
       String baseUri,
       String dockerImageName,
       String dockerImageTag,
-      CleanupStrategy cleanupStrategy) {
+      CleanupStrategy cleanupStrategy,
+      String containerName) {
     this.ignorePullError = ignorePullError;
     this.baseUri = baseUri;
     this.dockerImageName = dockerImageName;
     this.dockerImageTag = dockerImageTag;
     this.cleanupStrategy = cleanupStrategy;
+    this.containerName = containerName;
     this.gson = new Gson();
     this.requestFactory =
         new NetHttpTransport.Builder()
@@ -92,15 +97,17 @@ final class TestBench implements TestRule {
                 request -> {
                   request.setCurlLoggingEnabled(false);
                   request.getHeaders().setAccept("application/json");
-                  request.getHeaders().setUserAgent("test-bench/ java-conformance-tests/");
+                  request
+                      .getHeaders()
+                      .setUserAgent(String.format("%s/ test-bench/", this.containerName));
                 });
   }
 
-  String getBaseUri() {
+  public String getBaseUri() {
     return baseUri;
   }
 
-  RetryTestResource createRetryTest(RetryTestResource retryTestResource) throws IOException {
+  public RetryTestResource createRetryTest(RetryTestResource retryTestResource) throws IOException {
     GenericUrl url = new GenericUrl(baseUri + "/retry_test");
     String jsonString = gson.toJson(retryTestResource);
     HttpContent content =
@@ -112,14 +119,14 @@ final class TestBench implements TestRule {
     return result;
   }
 
-  void deleteRetryTest(RetryTestResource retryTestResource) throws IOException {
+  public void deleteRetryTest(RetryTestResource retryTestResource) throws IOException {
     GenericUrl url = new GenericUrl(baseUri + "/retry_test/" + retryTestResource.id);
     HttpRequest req = requestFactory.buildDeleteRequest(url);
     HttpResponse resp = req.execute();
     resp.disconnect();
   }
 
-  RetryTestResource getRetryTest(RetryTestResource retryTestResource) throws IOException {
+  public RetryTestResource getRetryTest(RetryTestResource retryTestResource) throws IOException {
     GenericUrl url = new GenericUrl(baseUri + "/retry_test/" + retryTestResource.id);
     HttpRequest req = requestFactory.buildGetRequest(url);
     HttpResponse resp = req.execute();
@@ -128,7 +135,7 @@ final class TestBench implements TestRule {
     return result;
   }
 
-  List<RetryTestResource> listRetryTests() throws IOException {
+  public List<RetryTestResource> listRetryTests() throws IOException {
     GenericUrl url = new GenericUrl(baseUri + "/retry_tests");
     HttpRequest req = requestFactory.buildGetRequest(url);
     HttpResponse resp = req.execute();
@@ -147,7 +154,8 @@ final class TestBench implements TestRule {
     return new Statement() {
       @Override
       public void evaluate() throws Throwable {
-        Path tempDirectory = Files.createTempDirectory("retry-conformance-server");
+        String fullContainerName = String.format("storage-testbench_%s", containerName);
+        Path tempDirectory = Files.createTempDirectory(fullContainerName);
         Path outPath = tempDirectory.resolve("stdout");
         Path errPath = tempDirectory.resolve("stderr");
 
@@ -190,7 +198,7 @@ final class TestBench implements TestRule {
                     "--rm",
                     "--publish",
                     port + ":9000",
-                    "--name=retry-conformance-server",
+                    String.format("--name=%s", fullContainerName),
                     dockerImage)
                 .redirectOutput(outFile)
                 .redirectError(errFile)
@@ -261,14 +269,31 @@ final class TestBench implements TestRule {
     }
   }
 
-  static Builder newBuilder() {
+  public static Builder newBuilder() {
     return new Builder();
   }
 
-  static final class RetryTestResource {
+  public static final class RetryTestResource {
     public String id;
     public Boolean completed;
     public JsonObject instructions;
+
+    public RetryTestResource() {}
+
+    public RetryTestResource(JsonObject instructions) {
+      this.instructions = instructions;
+    }
+
+    public static RetryTestResource newRetryTestResource(Method m, InstructionList l) {
+      RetryTestResource resource = new RetryTestResource();
+      resource.instructions = new JsonObject();
+      JsonArray instructions = new JsonArray();
+      for (String s : l.getInstructionsList()) {
+        instructions.add(s);
+      }
+      resource.instructions.add(m.getName(), instructions);
+      return resource;
+    }
 
     @Override
     public String toString() {
@@ -284,36 +309,46 @@ final class TestBench implements TestRule {
     }
   }
 
-  static final class Builder {
+  public static final class Builder {
     private static final String DEFAULT_BASE_URI = "http://localhost:9000";
     private static final String DEFAULT_IMAGE_NAME =
         "gcr.io/cloud-devrel-public-resources/storage-testbench";
-    private static final String DEFAULT_IMAGE_TAG = "v0.8.0";
+    private static final String DEFAULT_IMAGE_TAG = "v0.10.0";
+    private static final String DEFAULT_CONTAINER_NAME = "default";
 
     private boolean ignorePullError;
     private String baseUri;
     private String dockerImageName;
     private String dockerImageTag;
     private CleanupStrategy cleanupStrategy;
+    private String containerName;
 
-    public Builder() {
-      this(false, DEFAULT_BASE_URI, DEFAULT_IMAGE_NAME, DEFAULT_IMAGE_TAG, CleanupStrategy.ALWAYS);
+    private Builder() {
+      this(
+          false,
+          DEFAULT_BASE_URI,
+          DEFAULT_IMAGE_NAME,
+          DEFAULT_IMAGE_TAG,
+          CleanupStrategy.ALWAYS,
+          DEFAULT_CONTAINER_NAME);
     }
 
-    public Builder(
+    private Builder(
         boolean ignorePullError,
         String baseUri,
         String dockerImageName,
         String dockerImageTag,
-        CleanupStrategy cleanupStrategy) {
+        CleanupStrategy cleanupStrategy,
+        String containerName) {
       this.ignorePullError = ignorePullError;
       this.baseUri = baseUri;
       this.dockerImageName = dockerImageName;
       this.dockerImageTag = dockerImageTag;
       this.cleanupStrategy = cleanupStrategy;
+      this.containerName = containerName;
     }
 
-    public Builder setCleanupStraegy(CleanupStrategy cleanupStrategy) {
+    public Builder setCleanupStrategy(CleanupStrategy cleanupStrategy) {
       this.cleanupStrategy = requireNonNull(cleanupStrategy, "cleanupStrategy must be non null");
       return this;
     }
@@ -338,9 +373,19 @@ final class TestBench implements TestRule {
       return this;
     }
 
+    public Builder setContainerName(String containerName) {
+      this.containerName = requireNonNull(containerName, "containerName must be non null");
+      return this;
+    }
+
     public TestBench build() {
       return new TestBench(
-          ignorePullError, baseUri, dockerImageName, dockerImageTag, cleanupStrategy);
+          ignorePullError,
+          baseUri,
+          dockerImageName,
+          dockerImageTag,
+          cleanupStrategy,
+          containerName);
     }
   }
 }

--- a/google-cloud-storage/src/test/java/com/google/cloud/storage/it/ITBlobWriteChannelTest.java
+++ b/google-cloud-storage/src/test/java/com/google/cloud/storage/it/ITBlobWriteChannelTest.java
@@ -1,0 +1,194 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.storage.it;
+
+import static com.google.common.truth.Truth.assertThat;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotEquals;
+import static org.junit.Assert.assertTrue;
+
+import com.google.api.client.json.JsonParser;
+import com.google.api.gax.rpc.FixedHeaderProvider;
+import com.google.api.services.storage.model.StorageObject;
+import com.google.cloud.NoCredentials;
+import com.google.cloud.WriteChannel;
+import com.google.cloud.conformance.storage.v1.InstructionList;
+import com.google.cloud.conformance.storage.v1.Method;
+import com.google.cloud.storage.Blob;
+import com.google.cloud.storage.BlobInfo;
+import com.google.cloud.storage.BucketInfo;
+import com.google.cloud.storage.DataGeneration;
+import com.google.cloud.storage.PackagePrivateMethodWorkarounds;
+import com.google.cloud.storage.Storage;
+import com.google.cloud.storage.Storage.BlobWriteOption;
+import com.google.cloud.storage.StorageOptions;
+import com.google.cloud.storage.conformance.retry.TestBench;
+import com.google.cloud.storage.conformance.retry.TestBench.RetryTestResource;
+import com.google.cloud.storage.spi.v1.StorageRpc;
+import com.google.cloud.storage.testing.RemoteStorageHelper;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.reflect.Reflection;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Optional;
+import java.util.Random;
+import java.util.logging.Logger;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TestName;
+import org.threeten.bp.Clock;
+import org.threeten.bp.Instant;
+import org.threeten.bp.ZoneId;
+import org.threeten.bp.ZoneOffset;
+import org.threeten.bp.format.DateTimeFormatter;
+
+public final class ITBlobWriteChannelTest {
+  private static final Logger LOGGER = Logger.getLogger(ITBlobWriteChannelTest.class.getName());
+  private static final String NOW_STRING;
+
+  static {
+    Instant now = Clock.systemUTC().instant();
+    DateTimeFormatter formatter =
+        DateTimeFormatter.ISO_LOCAL_DATE_TIME.withZone(ZoneId.from(ZoneOffset.UTC));
+    NOW_STRING = formatter.format(now);
+  }
+
+  private static final String BUCKET = RemoteStorageHelper.generateBucketName();
+
+  @ClassRule public static final TestBench testBench = TestBench.newBuilder().build();
+
+  @Rule public final TestName testName = new TestName();
+
+  @Rule public final DataGeneration dataGeneration = new DataGeneration(new Random(1234567890));
+
+  /**
+   * Test for unexpected EOF at the beginning of trying to read the json response.
+   *
+   * <p>The error of this case shows up as an IllegalArgumentException rather than a json parsing
+   * error which comes from {@link JsonParser}{@code #startParsing()} which fails to find a node to
+   * start parsing.
+   */
+  @Test
+  public void testJsonEOF_0B() throws IOException {
+    int contentSize = 512 * 1024;
+    int cappedByteCount = 0;
+
+    doJsonUnexpectedEOFTest(contentSize, cappedByteCount);
+  }
+
+  /** Test for unexpected EOF 10 bytes into the json response */
+  @Test
+  public void testJsonEOF_10B() throws IOException {
+    int contentSize = 512 * 1024;
+    int cappedByteCount = 10;
+
+    doJsonUnexpectedEOFTest(contentSize, cappedByteCount);
+  }
+
+  private void doJsonUnexpectedEOFTest(int contentSize, int cappedByteCount) throws IOException {
+    String blobPath = String.format("%s/%s/blob", testName.getMethodName(), NOW_STRING);
+
+    BucketInfo bucketInfo = BucketInfo.of(BUCKET);
+    BlobInfo blobInfo = BlobInfo.newBuilder(bucketInfo, blobPath, 0L).build();
+
+    RetryTestResource retryTestResource =
+        RetryTestResource.newRetryTestResource(
+            Method.newBuilder().setName("storage.objects.insert").build(),
+            InstructionList.newBuilder()
+                .addInstructions(
+                    String.format("return-broken-stream-final-chunk-after-%dB", cappedByteCount))
+                .build());
+    RetryTestResource retryTest = testBench.createRetryTest(retryTestResource);
+
+    StorageOptions baseOptions =
+        StorageOptions.newBuilder()
+            .setCredentials(NoCredentials.getInstance())
+            .setHost(testBench.getBaseUri())
+            .setProjectId("project-id")
+            .build();
+    StorageRpc noHeader = (StorageRpc) baseOptions.getRpc();
+    StorageRpc yesHeader =
+        (StorageRpc)
+            baseOptions
+                .toBuilder()
+                .setHeaderProvider(
+                    FixedHeaderProvider.create(ImmutableMap.of("x-retry-test-id", retryTest.id)))
+                .build()
+                .getRpc();
+    //noinspection UnstableApiUsage
+    StorageOptions storageOptions =
+        baseOptions
+            .toBuilder()
+            .setServiceRpcFactory(
+                options ->
+                    Reflection.newProxy(
+                        StorageRpc.class,
+                        (proxy, method, args) -> {
+                          try {
+                            if ("writeWithResponse".equals(method.getName())) {
+                              boolean lastChunk = (boolean) args[5];
+                              LOGGER.info(
+                                  String.format(
+                                      "writeWithResponse called. (lastChunk = %b)", lastChunk));
+                              if (lastChunk) {
+                                return method.invoke(yesHeader, args);
+                              }
+                            }
+                            return method.invoke(noHeader, args);
+                          } catch (Exception e) {
+                            if (e.getCause() != null) {
+                              throw e.getCause();
+                            } else {
+                              throw e;
+                            }
+                          }
+                        }))
+            .build();
+
+    Storage testStorage = storageOptions.getService();
+
+    testStorage.create(bucketInfo);
+
+    ByteBuffer content = dataGeneration.randByteBuffer(contentSize);
+    // create a duplicate to preserve the initial offset and limit for assertion later
+    ByteBuffer expected = content.duplicate();
+
+    WriteChannel w = testStorage.writer(blobInfo, BlobWriteOption.generationMatch());
+    w.write(content);
+    w.close();
+
+    RetryTestResource postRunState = testBench.getRetryTest(retryTest);
+    assertTrue(postRunState.completed);
+
+    Optional<StorageObject> optionalStorageObject =
+        PackagePrivateMethodWorkarounds.maybeGetStorageObjectFunction().apply(w);
+
+    assertTrue(optionalStorageObject.isPresent());
+    StorageObject storageObject = optionalStorageObject.get();
+    assertThat(storageObject.getName()).isEqualTo(blobInfo.getName());
+
+    Blob blobGen2 = testStorage.get(blobInfo.getBlobId());
+    assertEquals(contentSize, (long) blobGen2.getSize());
+    assertNotEquals(blobInfo.getGeneration(), blobGen2.getGeneration());
+    ByteArrayOutputStream actualData = new ByteArrayOutputStream();
+    blobGen2.downloadTo(actualData);
+    ByteBuffer actual = ByteBuffer.wrap(actualData.toByteArray());
+    assertEquals(expected, actual);
+  }
+}

--- a/google-cloud-storage/src/test/java/com/google/cloud/storage/it/ITBlobWriteChannelTest.java
+++ b/google-cloud-storage/src/test/java/com/google/cloud/storage/it/ITBlobWriteChannelTest.java
@@ -71,7 +71,15 @@ public final class ITBlobWriteChannelTest {
 
   private static final String BUCKET = RemoteStorageHelper.generateBucketName();
 
-  @ClassRule public static final TestBench testBench = TestBench.newBuilder().build();
+  @ClassRule
+  public static final TestBench testBench =
+      TestBench.newBuilder()
+          .setContainerName("blob-write-channel-test")
+          // set a different base uri to prevent collision with conformance tests.
+          // It's possible for ITRetryConformanceTest to finish running and start shutting down
+          // when this test starts so the liveness check passes against the shutting down server.
+          .setBaseUri("http://localhost:9100")
+          .build();
 
   @Rule public final TestName testName = new TestName();
 

--- a/google-cloud-storage/src/test/java/com/google/cloud/storage/it/ITStorageTest.java
+++ b/google-cloud-storage/src/test/java/com/google/cloud/storage/it/ITStorageTest.java
@@ -67,6 +67,7 @@ import com.google.cloud.storage.BucketInfo.LifecycleRule.LifecycleAction;
 import com.google.cloud.storage.BucketInfo.LifecycleRule.LifecycleCondition;
 import com.google.cloud.storage.CopyWriter;
 import com.google.cloud.storage.Cors;
+import com.google.cloud.storage.DataGeneration;
 import com.google.cloud.storage.HmacKey;
 import com.google.cloud.storage.HttpMethod;
 import com.google.cloud.storage.PostPolicyV4;
@@ -116,7 +117,6 @@ import java.lang.reflect.Method;
 import java.net.URL;
 import java.net.URLConnection;
 import java.nio.ByteBuffer;
-import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.security.Key;
@@ -216,6 +216,7 @@ public class ITStorageTest {
       ImmutableList.of(LIFECYCLE_RULE_1, LIFECYCLE_RULE_2);
 
   @Rule public final TestName testName = new TestName();
+  @Rule public final DataGeneration dataGeneration = new DataGeneration(new Random(1234567890));
 
   @BeforeClass
   public static void beforeClass() throws IOException {
@@ -3860,13 +3861,13 @@ public class ITStorageTest {
     BlobId blobId = BlobId.of(BUCKET, blobPath);
     BlobInfo blobInfo = BlobInfo.newBuilder(blobId).build();
 
-    Random rand = new Random(1234567890);
-    String randString = randString(rand, contentSize);
-    final byte[] randStringBytes = randString.getBytes(StandardCharsets.UTF_8);
+    ByteBuffer contentGen1 = dataGeneration.randByteBuffer(contentSize);
+    ByteBuffer contentGen2 = dataGeneration.randByteBuffer(contentSize);
+    ByteBuffer contentGen2Expected = contentGen2.duplicate();
     Storage storage = StorageOptions.getDefaultInstance().getService();
     WriteChannel ww = storage.writer(blobInfo);
     ww.setChunkSize(chunkSize);
-    ww.write(ByteBuffer.wrap(randStringBytes));
+    ww.write(contentGen1);
     ww.close();
 
     Blob blobGen1 = storage.get(blobId);
@@ -3926,8 +3927,7 @@ public class ITStorageTest {
     try (WriteChannel w = testStorage.writer(blobGen1, BlobWriteOption.generationMatch())) {
       w.setChunkSize(chunkSize);
 
-      ByteBuffer buffer = ByteBuffer.wrap(randStringBytes);
-      w.write(buffer);
+      w.write(contentGen2);
     }
 
     assertTrue("Expected an exception to be thrown for the last chunk", exceptionThrown.get());
@@ -3937,18 +3937,6 @@ public class ITStorageTest {
     assertNotEquals(blobInfo.getGeneration(), blobGen2.getGeneration());
     ByteArrayOutputStream actualData = new ByteArrayOutputStream();
     blobGen2.downloadTo(actualData);
-    assertArrayEquals(randStringBytes, actualData.toByteArray());
-  }
-
-  private static String randString(Random rand, int length) {
-    final StringBuilder sb = new StringBuilder();
-    while (sb.length() < length) {
-      int i = rand.nextInt('z');
-      char c = (char) i;
-      if (Character.isLetter(c) || Character.isDigit(c)) {
-        sb.append(c);
-      }
-    }
-    return sb.toString();
+    assertEquals(contentGen2Expected, ByteBuffer.wrap(actualData.toByteArray()));
   }
 }

